### ISSUE TITLE
[BUG] more than 65535 masks in image, masks returned as np.uint32 . Segmentation striations. 

### ISSUE THREAD

**Issue by matthew-lee1** [MouseLand/cellpose#741]
**Describe the bug**
I am running cell segmentation on an image of size (53280, 29760). I am running this on CPU. I am using Cellpose version 2.2.2. When I run Cellpose, the only error that I receive is a warning which says: 
"more than 65535 masks in image, masks returned as np.uint32"
However, what I am finding is that there are only 65535 unique cells, and I am sure that there are more than this (somewhere between 400k - 600k cells). Looking at the source code I am not sure why it would be throwing this warning but not making the cast (which seems to be what is happening?)? 
Additionally as shown in the screenshot there are some weird striations on the mask, which do not appear in the flows visual. 

Any help would be greatly appreciated. Thanks!

Run Code:

from cellpose import models,core
from cellpose import plot,io
import time, os, sys, random
import numpy as np
import skimage.io
import shutil
import cv2
import matplotlib.pyplot as plt
import matplotlib as mpl
from urllib.parse import urlparse
from tifffile import imread, imsave, imwrite
from skimage.util import img_as_ubyte

use_GPU = core.use_gpu()
print('>>> GPU activated? %d'%use_GPU)

#from cellpose.io import logger_setup
#logger_setup();

##### SETTINGS 
#HIDING THE CODE IN THIS SECTION, JUST READING IN IMAGE WHICH OCCURS WITHOUT PROBLEM
f = input_dir + file

#Show 
print("Image location: ", f)
print("Output location: ", save_dir)

#Read in image and print dimensions
im = io.imread(f)
dim = im.shape
print(dim)

#define CHANNELS to run segementation on
#grayscale=0, R=1, G=2, B=3
#channels = [cytoplasm, nucleus]
#if NUCLEUS channel does not exist, set the second channel to 0
#channels = [0,0]
#IF ALL YOUR IMAGES ARE THE SAME TYPE, you can give a list with 2 elements
#channels = [0,0] # IF YOU HAVE GRAYSCALE
#channels = [2,3] # IF YOU HAVE G=cytoplasm and B=nucleus
#channels = [2,1] # IF YOU HAVE G=cytoplasm and R=nucleus
cell_channel = 1
nuclei_channel = 0
channels = [cell_channel, nuclei_channel]

diameter = 30
#print("Diameter is set to None. The size of the cells will be estimated on a per image basis")
    
#####DEFINE CELLPOSE MODEL
#model_type='cyto' or model_type='nuclei'
model_cell = models.Cellpose(gpu = use_GPU, model_type = 'cyto')
#model_nuclei = models.Cellpose(gpu = use_GPU, model_type = 'nuclei')

#####SEGMENTATION
print("Starting Segmentation")
masks_cell, flows_cell, styles_cell, diams_cell = model_cell.eval(im, 
                                                                  diameter=diameter,
                                                                  flow_threshold=0,
                                                                  channels=channels)

#Cell
mask_cell_output_name = save_dir + "MASK_CELL2_" + file_name + ".tif"
flow_cell_output_name = save_dir + "FLOWS_CELL2_" + file_name + ".tif"
print("Writing out mask")
masks_cell = masks_cell.astype(np.uint16) 
print("Writing out flows")
flows_cell_image = flows_cell[0].astype(np.uint8)

skimage.io.imsave(mask_cell_output_name, masks_cell, check_contrast = False)
skimage.io.imsave(flow_cell_output_name, flows_cell_image, check_contrast=False)

<img width="468" alt="image" src="https://github.com/MouseLand/cellpose/assets/91571913/6def879b-08ee-4cd5-95d3-09b65afba9f5">

**Comment by christianrickert** [comment]
Hi @matthew-lee1 

`cellpose` implements a [bounce check](https://github.com/MouseLand/cellpose/blob/869f16018ff82f0914c139a7eeb705a1da3c8da5/cellpose/io.py#L499-L512) that would've prevented your mask count limitation.

Instead, your code forcefully limits the number of available masks to 16 bits:
```python
model_cell = models.Cellpose(gpu = use_GPU, model_type = 'cyto')
masks_cell, flows_cell, styles_cell, diams_cell = model_cell.eval(im, diameter=diameter, flow_threshold=0, channels=channels)
mask_cell_output_name = save_dir + "MASK_CELL2_" + file_name + ".tif"
masks_cell = masks_cell.astype(np.uint16) # loss of information here
skimage.io.imsave(mask_cell_output_name, masks_cell, check_contrast = False)
```
1. Your value cast with [`numpy.ndarray.astype`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.astype.html) to an [`numpy.uint16`](https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.ushort) is `unsafe` by default, which means your mask values roll over from `65_535` to `0` repeatedly. You ran into a typical [overflow error](https://numpy.org/doc/stable/user/basics.types.html#overflow-errors).
2. The striations you see happen because the image display is scaled to display values from `0` to `65_535` and the image is produced from top (smaller mask values) to bottom (larger mask values) with repeated value "overflows". The more masks are present in a given segment (cell density), the faster the values roll over - with a sudden drop from maximum to minimum intensity. You're effectively looking at the [modulo](https://en.wikipedia.org/wiki/Modulo) of your cell id values, i.e. `cell_id % 65535 = mask_value`.
3. The image displays between `8 * 65_535 = 524_280` and `9 * 65_535 = 589_815` (unique) cell mask values in your example image. Casting your mask array to a `np.uint32` will suffice to resolve your issue.
4. Please read up on [`C data types`](https://en.wikipedia.org/wiki/C_data_types#Main_types) , which are the basis for `Python` data types.

My guess is that you ran into an error when saving values with more than 16 bits into a grayscale [`PNG`](https://en.wikipedia.org/wiki/PNG#Pixel_format) file. You then limited the mask values to 16 bits, but finally saved the same (limited) data into a `TIFF` instead.

In doubt, let Python handle the data types and restrict value ranges only when you are certain about the expected value range.

**Comment by rmd13** [comment]
May I ask if I use  io from cellpose to write output images, should I worry about this problem?


```
import numpy as np
from cellpose import models, io, utils
from cellpose.io import imread
import os
import csv
import datetime

masks, flows, styles, diameter = model.eval(image, diameter=diameterIn, channels=channels, flow_threshold=flow_threshold, cellprob_threshold=cellprob_threshold, stitch_threshold=stitch_threshold, do_3D=do_3D,anisotropy=anisotropy)
 
io.imsave(file_out,masks)
```
